File: datasets/C++/Chatgpt/Algorithm/hard5.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

const int MOD = 1e9 + 7;

// Compute XOR of a vector
int computeXOR(const vector<int>& arr) {
    int res = 0;
    for (int val : arr)
        res ^= val;
    return res;
}

int countNimGames(int n, int m) {
    int limit = (1 << m);  // 2^m
    vector<int> nums;

    // Only values from 1 to 2^m - 1
    for (int i = 1; i < limit; ++i)
        nums.push_back(i);

    int total = 0;
    vector<int> temp;

    // Generate all combinations of n values
    int size = nums.size();
    vector<bool> select(size, false);
    fill(select.end() - n, select.end(), true);

    do {
        temp.clear();
        for (int i = 0; i < size; ++i)
            if (select[i]) temp.push_back(nums[i]);

        // Generate all permutations of this combination
        sort(temp.begin(), temp.end());
        do {
            if (computeXOR(temp) != 0)
                total++;
        } while (next_permutation(temp.begin(), temp.end()));

    } while (next_permutation(select.begin(), select.end()));

    return total % MOD;
}

int main() {
    int n, m;
    cin >> n >> m;
    cout << countNimGames(n, m) << endl;
    return 0;
}

File: datasets/Python/Chatgpt/Algorithm/hard5.py
import sys
from itertools import combinations, permutations

MOD = 10**9 + 7

def xor(lst):
    res = 0
    for val in lst:
        res ^= val
    return res

def count_nim_games(n, m):
    nums = list(range(1, 1 << m))  # [1, 2, ..., 2^m - 1]
    total = 0
    for comb in combinations(nums, n):
        for perm in permutations(comb):
            if xor(perm) != 0:
                total += 1
    return total % MOD

if __name__ == "__main__":
    n, m = map(int, input().split())
    print(count_nim_games(n, m))
